[MRG+2] Implements ducktyping to allow for alternative Memory implementations

thechargedneutron opened the pull request:
#### Reference Issue
Fixes #9563 

#### What does this implement/fix? Explain your changes.
Introduced ducktyping in `pipeline.py` as well as in `hierarchical.py` 

#### Any other comments?
Added tests as well.

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

thechargedneutron committed:
Initial Commit

thechargedneutron committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into third

thechargedneutron commented:
How to import `joblib`. Tests failed because of ImportError. Do I need to mention somewhere that I used a new module?

jnothman commented on sklearn/cluster/hierarchical.py:
```diff
@@ -15,7 +15,7 @@
 
 from ..base import BaseEstimator, ClusterMixin
-from ..externals.joblib import Memory
+from joblib import Memory
```
I don't think this needs importing if all references to it in code have been removed...?

jnothman reviewed (commented):
I think in practice we also access `memory.cachedir`. Seeing as this is just used as a proxy for "is there actual caching happening", we can just use `getattr(memory, 'cachedir', True)` in its place.

thechargedneutron commented:
In the code segment 
```
elif isinstance(memory, six.string_types):
            memory = Memory(cachedir=memory, verbose=0)
```
we also check that if `memory` is passed as a string, we assign a Memory instance with `cachedir` as the string provided. Can we achieve it after removing the dependency on `joblib` ?

thechargedneutron committed:
Import removed in hierarchical

thechargedneutron committed:
Minor change

jnothman commented:
Ah, sorry, I made the wrong comment. There, we should still use
sklearn.externals.joblib.Memory.

What I think we should be doing here is adding a new helper in
sklearn.utils.validation, called check_memory, which would validate the
input, and return something with .cache. Then you only need the test there.

On 19 August 2017 at 23:08, Kumar Ashutosh <notifications@github.com> wrote:

> In the code segment
>
> elif isinstance(memory, six.string_types):
>             memory = Memory(cachedir=memory, verbose=0)
>
> we also check that if memory is passed as a string, we assign a Memory
> instance with cachedir as the string provided. Can we achieve it after
> removing the dependency on joblib ?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9584#issuecomment-323522181>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6-OqUxjt0M7NNbGEOXk7tV8EUzGjks5sZt40gaJpZM4O7xaJ>
> .
>

thechargedneutron commented:
I removed those lines and added a check memory.cachedir condition. Can you please check the last commit I made and decide whether it works?

thechargedneutron committed:
check_memory added

thechargedneutron committed:
Conflict Resolved

thechargedneutron committed:
Changes added

thechargedneutron committed:
Mistake removed

thechargedneutron committed:
Parameters revised

thechargedneutron committed:
test_pipeline revised

thechargedneutron committed:
cachedir added in Dummy()

thechargedneutron committed:
ValueError msg edited

thechargedneutron committed:
Pep8 errors removed

thechargedneutron committed:
Pep8 failures removed

thechargedneutron commented:
I hope this satisfies the requirements. Kindly review it and suggest changes, if any.

glemaitre commented:
You should change your title to MRG. You will get more attention from the developers :)

thechargedneutron commented:
I am waiting for a review. I think it's complete but not sure ;) What do you think?

glemaitre commented on sklearn/utils/validation.py:
```diff
@@ -155,6 +156,14 @@ def _shape_repr(shape):
 
 
+def check_memory(memory):
+    if memory is None:
```
You need to add a docstring

glemaitre commented on sklearn/utils/validation.py:
```diff
@@ -155,6 +156,14 @@ def _shape_repr(shape):
+        memory = Memory(cachedir=None, verbose=0)
+    elif isinstance(memory, six.string_types):
+        memory = Memory(cachedir=memory, verbose=0)
+    return memory
```
I think that the raise statement should come here.

glemaitre commented on sklearn/cluster/hierarchical.py:
```diff
@@ -693,14 +695,10 @@ def fit(self, X, y=None):
-            memory = Memory(cachedir=memory, verbose=0)
-        elif not isinstance(memory, Memory):
+        memory = check_memory(self.memory)
+        if not hasattr(memory, 'cache'):
```
This raise statement will be duplicate. You need to include it in `check_memory`.

glemaitre commented on sklearn/pipeline.py:
```diff
@@ -187,14 +188,10 @@ def _final_estimator(self):
-            memory = Memory(cachedir=memory, verbose=0)
-        elif not isinstance(memory, Memory):
+        memory = check_memory(self.memory)
+        if not hasattr(memory, 'cache'):
```
This raise statement will be duplicate. You need to include it in `check_memory`.

glemaitre commented on sklearn/cluster/tests/test_hierarchical.py:
```diff
@@ -140,6 +141,30 @@ def test_agglomerative_clustering_wrong_arg_memory():
     assert_raises(ValueError, clustering.fit, X)
 
 
+class Dummy(object):
```
Don't call this Dummy. It is not straightforward for something that it is not used to the code. `DummyMemory` is more informative

glemaitre replied:
I think that those two class should be moved into the testing of `check_memory` and could be imported here.

glemaitre commented on sklearn/cluster/tests/test_hierarchical.py:
```diff
@@ -140,6 +141,30 @@ def test_agglomerative_clustering_wrong_arg_memory():
+        return func
+
+
+class Wrong_Dummy(object):
```
don't use `_`. `WrongDummyMemory` is a better name.

glemaitre commented on sklearn/utils/validation.py:
```diff
@@ -155,6 +156,14 @@ def _shape_repr(shape):
     return "(%s)" % joined
 
 
+def check_memory(memory):
```
You need to test this functions in `test_validation.py`. Create the `DummyMemory` and `WrongDummyMemory` classes there.

glemaitre commented on sklearn/cluster/tests/test_hierarchical.py:
```diff
@@ -140,6 +141,30 @@ def test_agglomerative_clustering_wrong_arg_memory():
+    clustering.fit(X)
+
+    clustering = AgglomerativeClustering(memory=Wrong_Dummy())
+    assert_raises(ValueError, clustering.fit, X)
```
Use assert_raises_message or assert_raises_regex to check the message which is raised.

glemaitre commented on sklearn/tests/test_pipeline.py:
```diff
@@ -266,6 +266,31 @@ def test_pipeline_sample_weight_unsupported():
+
+    pipe = Pipeline([('transf', Transf()), ('clf', Mult())],
+                    memory=Wrong_Dummy())
+    assert_raises(ValueError, pipe.fit, X)
```
Use assert_raises_message or assert_raises_regex to check the message which is raised.

glemaitre commented on sklearn/tests/test_pipeline.py:
```diff
@@ -852,8 +877,10 @@ def test_pipeline_wrong_memory():
+    assert_raises_regex(ValueError, "'memory' should either be a string or"
+                        " a joblib.Memory"
+                        " instance, got 'memory={!r}' instead.".format(
+                                 type(memory)),
```
Hard code type(memory). We don't want the test to be able to modify depending of the type of memory.

glemaitre reviewed (commented):
You need to add `check_memory` in the documentation in `doc/modules/classes.rst` and `doc/developers/utilities.rst`
You need also to add an entry in the what's new.

jnothman commented:
> I am waiting for a review. I think it's complete but not sure ;) What do you think?

WIP means: Hello everyone this is a snippet of what I'm working on. WDYT?

MRG means: I've done what I'm going to do to make this mergeable. Now I need review.

thechargedneutron committed:
Changes added

thechargedneutron committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into third

thechargedneutron committed:
Import statements added

glemaitre commented on sklearn/cluster/tests/test_hierarchical.py:
```diff
@@ -34,6 +34,7 @@
 from sklearn.utils.fast_dict import IntFloatDict
 from sklearn.utils.testing import assert_array_equal
 from sklearn.utils.testing import assert_warns
+from sklearn.utils.tests.test_validation import DummyMemory, WrongDummyMemory
```
I would need @jnothman to tell me if doing this is fine.
I am not sure anymore that it is usually done in that way (even if we avoid code duplication)

jnothman replied:
TBH, once we're using `check_memory` it's not that important to have a test in `test_hierarchical.py`. Yes, it's good to have some kind of integration testing, but I don't think we get much value out of it in this case. I'd happily see this PR without the changes to test_hierarchical and test_pipeline.

jnothman commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -539,3 +540,23 @@ def test_suppress_validation():
+
+
+def test_check_memory():
+    memory = check_memory(DummyMemory())
```
we also need to test what happens when a string or None is passed in.

jnothman commented on sklearn/utils/validation.py:
```diff
@@ -155,6 +156,32 @@ def _shape_repr(shape):
+
+    Parameters
+    ----------
+    memory: input object.
```
need space before `:`. The right of the colon should only specify type, i.e. str or something with `cache` or None

jnothman commented on sklearn/utils/validation.py:
```diff
@@ -155,6 +156,32 @@ def _shape_repr(shape):
+
+    Returns
+    -------
+    memory: the input memory if it is valid. A valueError if invalid memory instance.
```
need space before `:`. The right of the colon should only specify type, i.e. Memory.

jnothman commented on sklearn/utils/validation.py:
```diff
@@ -155,6 +156,32 @@ def _shape_repr(shape):
+    """Check that the memory is an instance of joblib.Memory.
+
+    Raises a ValueError if the passed object does not have a
+
```
no blank line here, please

thechargedneutron committed:
Changes added

thechargedneutron committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into third

thechargedneutron commented:
Added the changes. Also removed the test from `test_hierarchical.py `. Also, since we are now testing memory separately, is `test_agglomerative_clustering_wrong_arg_memory` in `test_hierarchical.py ` required now?

glemaitre commented on sklearn/cluster/hierarchical.py:
```diff
@@ -609,7 +611,7 @@ class AgglomerativeClustering(BaseEstimator, ClusterMixin):
         If linkage is "ward", only "euclidean" is accepted.
 
-    memory : Instance of sklearn.externals.joblib.Memory or string, optional \
+    memory : Instance of joblib.Memory or string, optional \
```
Is it fitting on a single line?

glemaitre commented on sklearn/cluster/hierarchical.py:
```diff
@@ -779,7 +772,7 @@ class FeatureAgglomeration(AgglomerativeClustering, AgglomerationTransform):
         If linkage is "ward", only "euclidean" is accepted.
 
-    memory : Instance of sklearn.externals.joblib.Memory or string, optional \
+    memory : Instance of joblib.Memory or string, optional \
```
Is it fitting on a single line

glemaitre commented on sklearn/pipeline.py:
```diff
@@ -52,7 +53,7 @@ class Pipeline(_BaseComposition):
         an estimator.
 
-    memory : Instance of sklearn.external.joblib.Memory or string, optional \
+    memory : Instance of joblib.Memory or string, optional \
```
Is it fitting on one line

glemaitre commented on sklearn/pipeline.py:
```diff
@@ -538,7 +530,7 @@ def make_pipeline(*steps, **kwargs):
     *steps : list of estimators,
 
-    memory : Instance of sklearn.externals.joblib.Memory or string, optional \
+    memory : Instance of joblib.Memory or string, optional \
```
one line

glemaitre commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -539,3 +540,29 @@ def test_suppress_validation():
+    def cache(self, func):
+        return func
+
+    cachedir = None
```
put this in `__init__` method

jnothman replied:
We don't want to require cachedir, I think

glemaitre replied:
It could be useful for later checking maybe.

jnothman replied:
I'm not sure what you mean. An example?

What I mean is that we might as well accept (and therefore test) objects that have a `cache` method which operates as expected, regardless of whether it's got all the other attributes that a `joblib.Memory` has.

glemaitre replied:
oh ok.

I only wanted to make test something like 

```python
memory = check_memory(DummyMemory())
assert_equal(memory.cachedir = 'something')
```

but it can be cachedir or anything else as attributes

jnothman replied:
Not just:

```
dummy = DummyMemory()
memory = check_memory(dummy)
assert memory is dummy
```

glemaitre replied:
yes much better indeed

glemaitre commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -539,3 +540,29 @@ def test_suppress_validation():
+
+
+def test_check_memory():
+    try:
```
Don't use a try statement. Just make the `check_memory` call. If the error is raised it will be more informative.

glemaitre commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -539,3 +540,29 @@ def test_suppress_validation():
+
+def test_check_memory():
+    try:
+        check_memory("TestString")
```
put the results in a variable and check that the `cachedir` is set properly.

glemaitre commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -539,3 +540,29 @@ def test_suppress_validation():
+def test_check_memory():
+    try:
+        check_memory("TestString")
+        check_memory(None)
```
Make a similar test

glemaitre commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -539,3 +540,29 @@ def test_suppress_validation():
+    try:
+        check_memory("TestString")
+        check_memory(None)
+        check_memory(DummyMemory())
```
you might want to set cachedir to something different than None in DummyMemory. You can make a check then

glemaitre commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -539,3 +540,29 @@ def test_suppress_validation():
+        check_memory(DummyMemory())
+    except ValueError:
+        assert False, "check_memory failed with ValueError"
+    assert_raises(ValueError, check_memory, 1)
```
use assert_raises_regex to check the message raised

glemaitre commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -539,3 +540,29 @@ def test_suppress_validation():
+    except ValueError:
+        assert False, "check_memory failed with ValueError"
+    assert_raises(ValueError, check_memory, 1)
+    assert_raises(ValueError, check_memory, WrongDummyMemory())
```
use assert_raises_regex to check the message raised

glemaitre commented on sklearn/utils/validation.py:
```diff
@@ -155,6 +156,30 @@ def _shape_repr(shape):
 
+def check_memory(memory):
+    """Check that the memory is an instance of joblib.Memory.
+    Raises a ValueError if the passed object does not have a
```
You need a blank line between the short description and the long description.

glemaitre commented on sklearn/utils/validation.py:
```diff
@@ -155,6 +156,30 @@ def _shape_repr(shape):
+
+    Parameters
+    ----------
+    memory : string or None or something with cache.
```
"something with cache" is not really informative. 

```
... or user Memory implementing a cache method
```

glemaitre replied:
You also need a description

```
memory : string, None or user Memory implementing a cache method.
    If memory is a string, a joblib.Memory will be used with ``cachedir``corresponding to the provided string.
    If None is given, no caching is done and the Memory object is completely transparent.
    If a user-specified Memory, the instance need a ``cache`` method.

```

glemaitre commented on sklearn/utils/validation.py:
```diff
@@ -155,6 +156,30 @@ def _shape_repr(shape):
+
+    Returns
+    -------
+    memory : Memory
```
```
memory : Memory
    Validated memory instance.
```

glemaitre commented on sklearn/utils/validation.py:
```diff
@@ -155,6 +156,30 @@ def _shape_repr(shape):
+    elif isinstance(memory, six.string_types):
+        memory = Memory(cachedir=memory, verbose=0)
+    elif not hasattr(memory, 'cache'):
+            raise ValueError("'memory' should either be a string or"
```
I don't find the error message enough meaningful.

```
'memory' is not a string or a Memory instance implementing a cache method. Got a {} instance, instead.'.format(type(memory))
```

glemaitre reviewed (commented):
Add a whats new entry as well

thechargedneutron committed:
Changes added

thechargedneutron committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into third

thechargedneutron committed:
assert

thechargedneutron committed:
cachedir name changed

thechargedneutron committed:
cachedir format updated

thechargedneutron commented:
On assigning cachedir the value of passed string, `TestString` in my case. The build failure displays `TestString//joblib != TestString` and also `TestString/joblib != TestString` (in different build tests). Can one of you please point out the reason for this.
And also, where and what do I need to add in `whats_new.rst`. Please suggest.

glemaitre commented on sklearn/tests/test_pipeline.py:
```diff
@@ -852,9 +852,8 @@ def test_pipeline_wrong_memory():
-                        " sklearn.externals.joblib.Memory instance, got",
-                        cached_pipe.fit, X, y)
+    assert_raises_regex(ValueError, "'memory' is not a string or a Memory instance "
+                        "implementing a cache method. Got a {} instance, instead.".format(type(memory)), cached_pipe.fit, X, y)
```
PEP8

glemaitre replied:
79 characters limit

glemaitre commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -562,7 +562,7 @@ def __init__(self):
 
 def test_check_memory():
     memory = check_memory("TestString")
```
Don't use capital letter -> `"cache_directory"`

glemaitre commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -562,7 +562,7 @@ def __init__(self):
 def test_check_memory():
     memory = check_memory("TestString")
-    assert_equal(memory.cachedir, "TestString")
+    assert_equal(memory.cachedir, "TestString//joblib")
```
assert_equal(memory.cachedir, os.path.join('cache_dir", 'joblib')

glemaitre commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -10,7 +10,11 @@
 
 from sklearn.utils.testing import assert_true, assert_false, assert_equal
-from sklearn.utils.testing import assert_raises, assert_raises_regexp
+from sklearn.utils.testing import (
```
right it on 2 lines and don't import assert_raises_regexp because it is the same than assert_raises_regex. FInd and replace if you see it.
``` python 
from sklearn.utils.testing import assert_raises
from sklearn.utils.testing import assert_raises_regex
```

glemaitre commented on sklearn/tests/test_pipeline.py:
```diff
@@ -852,9 +852,8 @@ def test_pipeline_wrong_memory():
-    assert_raises_regex(ValueError, "'memory' should either be a string or a"
-                        " sklearn.externals.joblib.Memory instance, got",
-                        cached_pipe.fit, X, y)
+    assert_raises_regex(ValueError, "'memory' is not a string or a Memory instance "
```
PEP8

glemaitre commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -539,3 +545,34 @@ def test_suppress_validation():
+
+
+def test_check_memory():
+    memory = check_memory("TestString")
```
Don't use capital letter `"TestString"` -> `'cache_directory'`

glemaitre commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -539,3 +545,34 @@ def test_suppress_validation():
+
+def test_check_memory():
+    memory = check_memory("TestString")
+    assert_equal(memory.cachedir, "TestString//joblib")
```
assert_equal(memory.cachedir, os.join.path('cache_directory', 'joblib'))

glemaitre commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -539,3 +545,34 @@ def test_suppress_validation():
+                        "instead.".format(type(memory)), check_memory, 1)
+    assert_raises_regex(ValueError, "'memory' is not a string "
+                        "or a Memory instance implementing a cache method."
+                        " Got a {} instance, instead.".format(type(memory)),
```
as previously mentioned, do not use type(memory). state directly string in your pattern

glemaitre commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -539,3 +545,34 @@ def test_suppress_validation():
+    assert_raises_regex(ValueError, "'memory' is not a string "
+                        "or a Memory instance implementing a"
+                        " cache method. Got a {} instance, "
+                        "instead.".format(type(memory)), check_memory, 1)
```
state the int matching type directly

glemaitre commented on sklearn/utils/validation.py:
```diff
@@ -155,6 +156,38 @@ def _shape_repr(shape):
+    Returns
+    -------
+    memory : Memory
+        Validated memory instance
```
put a final full stop.

glemaitre reviewed (commented):
Almost there :)

This is a bug fix allowing to pass user-specific Memory using duck-typing. state also that a utility helper function `check_memory` has been created to validate the memory instance.

You can check the format use when doing bug fix in `doc/whats_new.rst`

thechargedneutron committed:
Added description in whats_new.rst and other fixes

thechargedneutron committed:
missing import added

jnothman commented on doc/modules/classes.rst:
```diff
@@ -1380,6 +1380,7 @@ Low-level methods
    utils.validation.check_symmetric
    utils.validation.column_or_1d
    utils.validation.has_fit_parameter
+   utils.validation.check_memory
```
please put in alphabetical order.

jnothman commented on doc/whats_new.rst:
```diff
@@ -59,6 +59,11 @@ Decomposition, manifold learning and clustering
   :user:`James Bourbeau <jrbourbeau>`.
 
+- Fixed a bug allowing user-specific Memory using duck-typing. Function
+  `utils.validation.check_memory` has been created to validate the
```
put `:func:` before the first backtick

jnothman commented on doc/whats_new.rst:
```diff
@@ -59,6 +59,11 @@ Decomposition, manifold learning and clustering
 
+- Fixed a bug allowing user-specific Memory using duck-typing. Function
+  `utils.validation.check_memory` has been created to validate the
+  memory instance. :issue: `9563` by
```
no space between `:issue:` and ` please.

jnothman commented on sklearn/utils/validation.py:
```diff
@@ -155,6 +156,38 @@ def _shape_repr(shape):
+
+    Parameters
+    ----------
+    memory : string or None or user Memory implementing a cache method.
```
just say string or None or Memory instance

jnothman commented on sklearn/utils/validation.py:
```diff
@@ -155,6 +156,38 @@ def _shape_repr(shape):
+        ``cachedir``corresponding to the provided string.
+        If None is given, no caching is done and the Memory
+        object is completely transparent.
+        If a user-specified Memory, the instance need a ``cache`` method.
```
Just say "Any object with a cache method will be accepted as a Memory instance"

jnothman reviewed (commented):
Good work!

So maybe I was wrong that we shouldn't test this in a Pipeline :\ It still tries to get `memory.cachedir`. Instead it should be doing `getattr(memory, 'cachedir', True)` in that context.

thechargedneutron committed:
Adds the suggested changes

thechargedneutron committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into third

thechargedneutron committed:
Error message revised

thechargedneutron commented:
I am confused between `<type 'int'> ` and `<class 'int'> `. When I run it locally,  i.e. check_memory(1) it raises an error having `<type 'int'>` in it, but in the build it says `<class 'int'>` . Please suggest what should I be writing there.
I hope the rest is fine.

jnothman commented:
The difference is between python 2 and 3. use a regular expression to allow
either

On 23 Aug 2017 6:07 pm, "Kumar Ashutosh" <notifications@github.com> wrote:

> I am confused between <type 'int'> and <class 'int'>. When I run it
> locally, i.e. check_memory(1) it raises an error having <type 'int'> in
> it, but in the build it says <class 'int'> . Please suggest what should I
> be writing there.
> I hope the rest is fine.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9584#issuecomment-324254231>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz61w2KwQBTsfCekEnJ5OxkUpUnpDkks5sa93YgaJpZM4O7xaJ>
> .
>

thechargedneutron committed:
changes added

thechargedneutron committed:
changed

thechargedneutron commented:
Kindly review it. Hope this works.

glemaitre commented on sklearn/cluster/hierarchical.py:
```diff
@@ -26,6 +25,8 @@
 
 from ..externals.six.moves import xrange
 
+from sklearn.utils.validation import check_memory
```
we usually make relative import:

```python
..utils.validation import check_memory
```

glemaitre replied:
Move this to line 21 as well under the other utils

glemaitre commented on sklearn/cluster/hierarchical.py:
```diff
@@ -196,7 +197,8 @@ def ward_tree(X, connectivity=None, n_clusters=None, return_distance=False):
         if n_clusters > n_samples:
             raise ValueError('Cannot provide more clusters than samples. '
-                             '%i n_clusters was asked, and there are %i samples.'
+                             '%i n_clusters was asked, and there are'
```
just to be sure. was it not PEP8 previosuly. If it was less than 80 characters avoid to make a diff for nothing.

glemaitre commented on sklearn/pipeline.py:
```diff
@@ -21,6 +21,7 @@
 from .utils import Bunch
 
 from .utils.metaestimators import _BaseComposition
+from sklearn.utils.validation import check_memory
```
make relative import as previously stated

glemaitre commented on sklearn/tests/test_pipeline.py:
```diff
@@ -868,9 +868,10 @@ def test_pipeline_wrong_memory():
+    assert_raises_regex(ValueError, "'memory' is not a string or a "
+                        "Memory instance implementing a cache method."
+                        " Got a {} instance, instead."
+                        .format(type(memory)), cached_pipe.fit, X, y)
```
Still do not make a `type` checking. Apply the same than what you did in the previous commit.

thechargedneutron replied:
I did not understand what I should have done here. Can you please elaborate.

glemaitre replied:
```python
assert_raises_regex(ValueError, "'memory' is not a string or a "
                    "Memory instance implementing a cache method."
                    " Got a *.int*. instance, instead.", cached_pipe.fit, X, y)
```

glemaitre commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -31,6 +33,7 @@
     check_is_fitted,
     check_consistent_length,
     assert_all_finite,
+    check_memory
```
Just as a note for you future PR:
We only make put absolute path in the test file like here :)

glemaitre reviewed (commented):
You need also to add an entry in http://scikit-learn.org/stable/developers/utilities.html#validation-tools

by editing 

https://github.com/scikit-learn/scikit-learn/blob/master/doc/developers/utilities.rst

```
- :func:`check_memory`: check that memory is a valid input. If a string is given, a ``joblib.Memory`` will be created. A user-specific Memory object having a ``cache`` method can be passed as well.
```

glemaitre commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -539,3 +543,35 @@ def test_suppress_validation():
+    assert memory is dummy
+    assert_raises_regex(ValueError, "'memory' is not a string "
+                        "or a Memory instance implementing a"
+                        " cache method. Got a " + str(type(1)) + " instance, "
```
I actually prefer the solution of @jnothman 

```python
assert_raises_regex(ValueError, "'memory' is not a string or a Memory"
                    " instance implementing a cache method. Got a "
                    " .*int.* instance, instead.", check_memory, 1)
```

thechargedneutron committed:
Changes added

thechargedneutron commented:
@glemaitre  Kindly review this commit.

glemaitre commented:
@thechargedneutron You still need to address 4 comments of my last review and add in the doc the utility as mentioned https://github.com/scikit-learn/scikit-learn/pull/9584#pullrequestreview-58027039

thechargedneutron committed:
Chenges added

thechargedneutron commented:
Really sorry. I forgot to git add before commit. Review this. I am sorry :(

glemaitre commented on sklearn/tests/test_pipeline.py:
```diff
@@ -868,8 +868,9 @@ def test_pipeline_wrong_memory():
-                        " sklearn.externals.joblib.Memory instance, got",
+    assert_raises_regex(ValueError, "'memory' is not a string or a "
+                        "Memory instance implementing a cache method."
+                        " Got a *.int*. instance, instead.",
```
`*.int*.` -> `.*int.*`

glemaitre replied:
That's why you have an error

thechargedneutron committed:
Error removed

glemaitre commented on doc/developers/utilities.rst:
```diff
@@ -177,6 +177,10 @@ Benchmarking
 Testing Functions
 =================
 
+- :func:`check_memory`: check that memory is a valid input. If a string is
```
Move to line 45. It is fitting better with a small descriptive paragraph instead of an item:

```
Some estimators are able to store intermediate results using ``Memory``
instance. The func:`check_memory`: check that memory is a valid input.
If a string is given, a ``joblib.Memory`` will be created. A user-specific 
``Memory`` object having a ``cache`` method can be passed as well.
```

glemaitre commented:
Just the doc comments and then @jnothman can make a last proof reading.
I think that that's all on my side.

thechargedneutron committed:
doc changed

glemaitre commented on doc/developers/utilities.rst:
```diff
@@ -43,6 +43,11 @@ should be used when applicable.
   be sliced or indexed using safe_index.  This is used to validate input for
   cross-validation.
 
+- Some estimators are able to store intermediate results using ``Memory``
```
Remove `-`

glemaitre reviewed (commented):
Apart of that LGTM

thechargedneutron committed:
- removed

thechargedneutron commented:
@jnothman Kindly review and suggest how to solve the codecov/patch failure.

glemaitre commented:
You need to cover the missing line if you want the patch to pass:
https://codecov.io/gh/scikit-learn/scikit-learn/compare/aae87002b96622424a16dcad2eaef3a75cc0feda...b86dc4ce2f9a592ae9550f488867122225982c02/src/sklearn/utils/tests/test_validation.py#L553

IMO, this is pretty useless since that we don't want to check the functionality of this dummy.
Basically you would need to do something like:

```python
transformer = AgglomerativeClustering(memory=DummyMemory())
transformer.fit_transform(X)
``` 

Like this the function cache of DummyMemory will be called. I am not sure this is useful (apart of making the line green).
I let @jnothman decide what the best

jnothman reviewed (commented):
I don't mind if that line doesn't get coverage, except that perhaps we should go back to testing pipeline with the dummy: it won't work currently.

thechargedneutron committed:
Tests again added in test_pipeline.py

thechargedneutron committed:
getattr added

thechargedneutron commented:
changing `if memory.cachedir is None:` with `if getattr(memory, 'cachedir', True): ` raises other errors while build. Do you think using both the conditions with an `or` will work?

thechargedneutron commented:
Using `or` is a bad idea. Try except may work. WDYT?

thechargedneutron committed:
setattr added to set cachedir to None by default

thechargedneutron committed:
changes added

thechargedneutron committed:
not statement removed

thechargedneutron committed:
Update pipeline.py

thechargedneutron commented:
I hope this works!!

thechargedneutron committed:
Merge branch 'third' of https://github.com/thechargedneutron/scikit-learn into third

jnothman commented on sklearn/cluster/hierarchical.py:
```diff
@@ -195,8 +196,8 @@ def ward_tree(X, connectivity=None, n_clusters=None, return_distance=False):
         if n_clusters > n_samples:
-            raise ValueError('Cannot provide more clusters than samples. '
-                             '%i n_clusters was asked, and there are %i samples.'
+            raise ValueError('Cannot provide more clusters than samples. %i '
```
Please do not change unrelated things. It makes your contribution harder to review and may introduce merge conflicts to other pull requests.

thechargedneutron replied:
Got it. Sorry for the inconveniences.

jnothman reviewed (commented):
This LGTM

glemaitre commented:
I see that @GaelVaroquaux is around in GitHub. Could you have a look for a potential MRG+2?

thechargedneutron commented:
@glemaitre @jnothman  Thanks for guiding me through!! :)

glemaitre commented:
@lesteve are the right man for a MRG + 2

lesteve commented on sklearn/utils/validation.py:
```diff
@@ -155,6 +156,34 @@ def _shape_repr(shape):
 
 
+def check_memory(memory):
+    """Check that the memory is an instance of joblib.Memory.
```
I think we need to come up with a better wording here and everywhere else in this PR. `joblib.Memory` is misleading since `joblib` may not be installed.

I would use "memory-like" (maybe Memory-like i.e.with a capital M), similarly to array-like in numpy.

The docstring would read something like

```py
"""Check that ``memory`` is memory-like.

memory-like means that ``memory`` can be converted into a
sklearn.externals.joblib.Memory instance (typically a str denoting the
``cachedir``) or has the same interface.

Parameters
----------
memory : memory-like or None

Returns
------
memory : object with the sklearn.externals.joblib.Memory interface

Raises
------
ValueError
    if ``memory`` is not memory-like
"""
```

And then use memory-like as the type info in all the docstrings, possibly with a rst link to check_memory for more details.

lesteve commented on sklearn/tests/test_pipeline.py:
```diff
@@ -868,11 +868,35 @@ def test_pipeline_wrong_memory():
 
 
+class DummyMemory(object):
+    def __init__(self):
```
You do not need to implement `__init__` just use

```py
class DummyMemory(object):
    def cache(self, func):
        return func
```

lesteve commented on sklearn/tests/test_pipeline.py:
```diff
@@ -868,11 +868,35 @@ def test_pipeline_wrong_memory():
+
+
+class WrongDummyMemory(object):
+    def __init__(self):
```
You don't need to implement `__init__`

lesteve commented on sklearn/pipeline.py:
```diff
@@ -209,7 +200,7 @@ def _fit(self, X, y=None, **fit_params):
                 pass
             else:
-                if memory.cachedir is None:
+                if getattr(memory, 'cachedir', True) is None:
```
The default value `True` is a bit magical/misleading here ... maybe one of the (rare) cases where `hasattr` is slightly easier to read:

```py
if hasattr(memory, 'cachedir') and memory.cachedir is None:
```

lesteve reviewed (commented):
First pass at this PR

thechargedneutron commented on sklearn/utils/validation.py:
```diff
@@ -155,6 +156,34 @@ def _shape_repr(shape):
+    elif isinstance(memory, six.string_types):
+        memory = Memory(cachedir=memory, verbose=0)
+    elif not hasattr(memory, 'cache'):
+            raise ValueError("'memory' is not a string or a Memory instance "
```
Can you please give an example? I do not understand what to print in case `memory` is a class.

thechargedneutron replied:
@glemaitre Can you please give an example of how to achieve this?

glemaitre replied:
I think that what @lesteve is mentioning is to `print(memory)` instead of `print(type(memory))`
In short put the value of memory instead of the type of memory.

thechargedneutron committed:
changes added

thechargedneutron committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into third

glemaitre commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -539,3 +543,35 @@ def test_suppress_validation():
+
+
+class WrongDummyMemory(object):
+    def __init__(self):
```
Don't need init as well

glemaitre commented on sklearn/utils/tests/test_validation.py:
```diff
@@ -539,3 +543,35 @@ def test_suppress_validation():
+
+
+class DummyMemory(object):
+    def __init__(self):
```
Don't need init as well

thechargedneutron replied:
Added.

GaelVaroquaux commented:
I am not sure that now is the time to do ducktyping for Memory: we are
considering changing the API of joblib.

amueller commented:
@GaelVaroquaux how do you suppose to fix the bug if someone passes a ``joblib.Memory`` object instead of a ``sklearn.externals.joblib.Memory`` then?

thechargedneutron committed:
__init__ removed

thechargedneutron committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into third

thechargedneutron committed:
prints memory instead of type

thechargedneutron commented:
@glemaitre  @lesteve  Added the changes suggested by you.

glemaitre commented:
@thechargedneutron thanks. Now this is on the hand of @lesteve @GaelVaroquaux  which have the big picture regarding the `joblib` API.

thechargedneutron committed:
changes added

thechargedneutron committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into third

lesteve commented:
> @GaelVaroquaux how do you suppose to fix the bug if someone passes a joblib.Memory object instead of a sklearn.externals.joblib.Memory then?

This is roughly what nilearn does from [here](https://github.com/nilearn/nilearn/blob/6fe6245f75c0440e551b6402874146fffa4bed69/nilearn/_utils/cache_mixin.py#L22):

```py
from sklearn.externals.joblib import Memory

MEMORY_CLASSES = (Memory, )

try:
    from joblib import Memory as JoblibMemory
    MEMORY_CLASSES = (Memory, JoblibMemory)
except ImportError:
    pass

if isinsance(memory, MEMORY_CLASSES):
    ...
```

I guess duck-typing is supposed to be more Pythonic but I would say it is rather unlikely that any other implementation of `Memory` exists in the wild so maybe the nilearn approach is fine. Maybe duck-typing has a slight edge for me because testing is more straightforward.

> I am not sure that now is the time to do ducktyping for Memory: we are considering changing the API of joblib.

@GaelVaroquaux the only requirement is that the `Memory` object has a cache attribute which seems quite minimal but maybe there are things in your head that I don't know about yet ;-).

thechargedneutron committed:
error message corrected

lesteve commented:
@thechargedneutron could you try to fix the tests in the mean time?

thechargedneutron commented:
@lesteve  Done!! Please review if there's something left unattended.

thechargedneutron committed:
Space removed

thechargedneutron committed:
Change added

thechargedneutron committed:
dummy variable added

thechargedneutron commented:
@lesteve  Tests fixed!! I used a dummy variable in order to match the object in the two instances. Please check if this would work.

lesteve commented on doc/developers/utilities.rst:
```diff
@@ -43,6 +43,11 @@ should be used when applicable.
   cross-validation.
 
+Some estimators are able to store intermediate results using ``Memory``
+instance. The func:`check_memory`: check that memory is a valid input.
```
problem with the rst syntax (missing colon at the beginning) this should be 
```
:func:`check_memory`
```

Please check other potential problems in the doc generated by CircleCI on your PR: https://13139-843222-gh.circle-artifacts.com/0/home/ubuntu/scikit-learn/doc/_build/html/stable/developers/utilities.html

lesteve replied:
I would simplify the wording and replace your paragraph with something like:
```
The :func:`check_memory` checks that input is ``joblib.Memory``-like,
which means that it can be converted into a
``sklearn.externals.joblib.Memory`` instance (typically a str denoting
the ``cachedir``) or has the same interface.
```

lesteve commented on doc/whats_new.rst:
```diff
@@ -63,6 +63,11 @@ Decomposition, manifold learning and clustering
   division on Python 2 versions. :issue:`9492` by
   :user:`James Bourbeau <jrbourbeau>`.
 
+- Fixed a bug allowing user-specific Memory using duck-typing.
```
You can be a little bit less verbose, i.e. something like:
```
Add  :func:`utils.validation.check_memory` to validate that input is ``joblib.Memory``-like. issue:`9563` by
:user:`Kumar Ashutosh <thechargedneutron>`
```

lesteve commented on sklearn/cluster/hierarchical.py:
```diff
@@ -609,8 +610,7 @@ class AgglomerativeClustering(BaseEstimator, ClusterMixin):
 
-    memory : Instance of sklearn.externals.joblib.Memory or string, optional \
-            (default=None)
+    memory : Instance of joblib.Memory or string, optional (default=None)
```
For type info, I would use joblib.Memory-like everywhere, in this case I would do:
```
memory : joblib.Memory-like, optional (default None)
    # same description as before goes here
```

lesteve commented on sklearn/cluster/hierarchical.py:
```diff
@@ -779,8 +770,7 @@ class FeatureAgglomeration(AgglomerativeClustering, AgglomerationTransform):
 
-    memory : Instance of sklearn.externals.joblib.Memory or string, optional \
-            (default=None)
+    memory : Instance of joblib.Memory or string, optional (default=None)
```
Use joblib.Memory-like here too.

lesteve commented on sklearn/pipeline.py:
```diff
@@ -537,8 +528,7 @@ def make_pipeline(*steps, **kwargs):
 
-    memory : Instance of sklearn.externals.joblib.Memory or string, optional \
-            (default=None)
+    memory : Instance of joblib.Memory or string, optional (default=None)
```
joblib.Memory-like here too.

lesteve commented on sklearn/utils/validation.py:
```diff
@@ -155,6 +156,38 @@ def _shape_repr(shape):
+    elif isinstance(memory, six.string_types):
+        memory = Memory(cachedir=memory, verbose=0)
+    elif not hasattr(memory, 'cache'):
+        raise ValueError("'memory' is not a string or a Memory instance "
```
Improve wording:
```py
"'memory' should be None, a string or have the same interface as sklearn.externals.joblib.Memory.
Got {} instead.
```

That means you need to amend the tests accordingly (if you change the error message).

lesteve commented on sklearn/pipeline.py:
```diff
@@ -51,8 +52,7 @@ class Pipeline(_BaseComposition):
 
-    memory : Instance of sklearn.external.joblib.Memory or string, optional \
-            (default=None)
+    memory : Instance of joblib.Memory or string, optional (default=None)
```
Use joblib.Memory-like here too.

lesteve commented on sklearn/utils/validation.py:
```diff
@@ -155,6 +156,38 @@ def _shape_repr(shape):
 
 
+def check_memory(memory):
+    """Check that the ``memory`` is Memory-like.
```
Remove the "the" and use joblib.Memory-like for consistency: "Check that ``memory`` is joblib.Memory-like"

lesteve commented on sklearn/utils/validation.py:
```diff
@@ -155,6 +156,38 @@ def _shape_repr(shape):
+def check_memory(memory):
+    """Check that the ``memory`` is Memory-like.
+
+    Memory-like means that ``memory`` can be converted into a
```
joblib.Memory-like

lesteve commented on sklearn/utils/validation.py:
```diff
@@ -155,6 +156,38 @@ def _shape_repr(shape):
+
+    Parameters
+    ----------
+    memory : Memory-like or None
```
joblib.Memory-like

lesteve commented on sklearn/utils/validation.py:
```diff
@@ -155,6 +156,38 @@ def _shape_repr(shape):
+
+    Returns
+    -------
+    memory : object with the sklearn.externals.joblib.Memory interface
```
with the joblib.Memory interface

lesteve commented on sklearn/utils/validation.py:
```diff
@@ -155,6 +156,38 @@ def _shape_repr(shape):
+    Raises
+    ------
+    ValueError
+        if ``memory`` is not Memory-like
```
joblib.Memory-like

lesteve commented on sklearn/utils/validation.py:
```diff
@@ -155,6 +156,38 @@ def _shape_repr(shape):
+        if ``memory`` is not Memory-like
+    """
+
+    if memory is None:
```
You can actually tackle this case in the same way as a string:

```py
if memory is None or isinstance(memory, six.string_types):
    memory = Memory(cachedir=memory, verbose=0)t
elif not hasattr(memory, 'cache'):
    # rest of the code is the same
```

lesteve reviewed (commented):
I am of the opinion of using joblib.Memory-like consistently. I have a few other comments as well.

thechargedneutron committed:
Changes added

thechargedneutron committed:
test_pipeline.py updated

thechargedneutron committed:
documentation revised

thechargedneutron commented:
@lesteve Changes added.

amueller commented:
@lesteve @GaelVaroquaux what if some other library starts to become a package-manager like scikit-learn and vendors joblib and the user is using the memory from that library? ;)

thechargedneutron committed:
parameter revised

GaelVaroquaux commented:
> @lesteve @GaelVaroquaux what if some other library starts to become a
> package-manager like scikit-learn and vendors joblib and the user is
> using the memory from that library? ;)

We'll do a leftpad on them :).

Anyhow, you convinced me that the PR needs to go in, and we'll cope with
the API change later (it's still WIP anyhow).

lesteve commented on sklearn/utils/validation.py:
```diff
@@ -155,6 +156,36 @@ def _shape_repr(shape):
+    elif not hasattr(memory, 'cache'):
+        raise ValueError("'memory' should be None, a string or have the same"
+                         " interface as sklearn.externals.joblib.Memory."
+                         " Got {} instead.".format(memory))
```
I would use "Got memory='{}' instead." for better readability. You will need to update the tests accordingly.

thechargedneutron replied:
Done.

lesteve reviewed (commented):
Just a minor comment, other than this LGTM.

thechargedneutron committed:
error statement revised

thechargedneutron committed:
pep8 failure removed

thechargedneutron commented:
@lesteve  @amueller  Changes added, tests complete. Please review.

lesteve commented:
LGTM. @jnothman and/or @glemaitre maybe you want to have a look?

glemaitre commented:
I'll make a review now

glemaitre commented on sklearn/cluster/hierarchical.py:
```diff
@@ -15,17 +15,18 @@
 
 from ..externals.six.moves import xrange
 
+
```
unnecessary space

glemaitre commented on sklearn/utils/validation.py:
```diff
@@ -155,6 +156,36 @@ def _shape_repr(shape):
+    Parameters
+    ----------
+    memory : joblib.Memory-like
+
```
@lesteve we should put a description to follow numpydoc? (That's true that there is no much to say apart of "memory instance to be validated" and "Validated memory instance"

glemaitre commented on sklearn/utils/validation.py:
```diff
@@ -155,6 +156,36 @@ def _shape_repr(shape):
+    Raises
+    ------
+    ValueError
+        if ``memory`` is not joblib.Memory-like
```
Capital letter and final full mark:

```
       If ``memory`` is not joblib.Memory-like.
```

glemaitre reviewed (commented):
LGTM, only nitpick ... @lesteve I would almost do them if I would have push rights :)

thechargedneutron committed:
edited

thechargedneutron committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into third

thechargedneutron commented:
@glemaitre  Done! except the 2nd comment.

glemaitre commented:
> @glemaitre Done! except the 2nd comment.

Great. Was it a review to remove them previously. If yes, I am fine for merging.

jnothman commented on sklearn/pipeline.py:
```diff
@@ -537,8 +528,7 @@ def make_pipeline(*steps, **kwargs):
 
-    memory : Instance of sklearn.externals.joblib.Memory or string, optional \
-            (default=None)
+    memory : joblib.Memory-like, optional (default=None)
```
I think we need to say string is an option in the type spec or we make this seem harder to use than it is. If we're short of space, I don't feel default=None adds anything.

thechargedneutron replied:
@lesteve  suggested this to me, WDYT should be here?

glemaitre replied:
Agreed, did not spot it.

thechargedneutron replied:
Should I add `or string` ?

glemaitre replied:
drop (default=None) for sure. I would add or string since it would work if you pass one.

thechargedneutron replied:
What's the reason for dropping `default=None` ? We are not short of space.

jnothman replied:
It's something we're not very consistent with here. I find that `default=None` is very uninformative:
1. Default values (but not their meaning) are mentioned in the function signature, and key projects like numpy etc do not mention them here in the type spec.
2. "optional" and "default=..." are redundant together.
3. "default=None" has a different semantics (i.e. it does something different) everywhere you see it: sometimes it means "determine it automatically"; sometimes "switch the feature off"; sometimes "all"; sometimes "none".

So I much prefer not to say "default=None", where "optional" already implies (and can be made explicit in the description) that the user can live without the feature.

But you certainly should state that passing a string is one way to specify `memory`.

jnothman commented:
I've marked this for inclusion in 0.19.1 because it may help a lot with the new feature in Pipeline.

thechargedneutron committed:
doc changed

thechargedneutron committed:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into third

thechargedneutron commented:
@jnothman  Changed the function description. Review please.

jnothman commented on sklearn/utils/validation.py:
```diff
@@ -155,6 +156,36 @@ def _shape_repr(shape):
+
+    Parameters
+    ----------
+    memory : joblib.Memory-like
```
or string or None

jnothman commented:
Fix that up and I'll merge on green.

thechargedneutron committed:
changes added

jnothman committed:
Formatting

jnothman committed:
save what's new for 0.19.1

jnothman committed:
Wording

jnothman commented:
Thanks @thechargedneutron!

thechargedneutron commented:
@jnothman welcome!! Thanks for the help :)

lesteve commented:
I am not sure why we added string on top of joblib.Memory-like explicitly to the type info.

joblib.Memory-like means that ``memory`` can be converted into a
sklearn.externals.joblib.Memory instance (typically a str denoting the
``cachedir``) or has the same interface (has a ``cache`` method).

In my mind, joblib.Memory-like was supposed to be modelled after array-like in numpy. You don't see  this in numpy docstrings:
```
parameter : array_like or list
```

I would agree that being explicit and adding string makes it easier to use for user not familiar with joblib.Memory (in general the docstring does say what happens if a string is passed in though). I was hoping that just using `joblib.Memory-like` in the type info and either adding a link to check_memory function for more details or copy and pasting the same docstring around would work.

jnothman commented:
Well, we can debate it, but I strongly disagree that "joblib.Memory-like"
is sufficiently helpful to a user who's never heard of joblib. If they're
told they can use a string, then they might actually consider using the
feature rather than saying "sounds like too much work".

Besides, I find the analogy list:array :: string:Memory to be quite
far-fetched. A some lists can be, essentially, cast into an array (the data
is copied into a different format in memory and given a different API), but
a string is not being cast as a Memory, it is being used as the primary
attribute that defines a Memory. I get that there are some parallels, but I
find the notion that a string is Memory-like to be unintuitive.

On 30 August 2017 at 16:11, Loïc Estève <notifications@github.com> wrote:

> I am not sure why we added string on top of joblib.Memory-like explicitly
> to the type info.
>
> joblib.Memory-like means that memory can be converted into a
> sklearn.externals.joblib.Memory instance (typically a str denoting the
> cachedir) or has the same interface (has a cache method).
>
> In my mind, joblib.Memory-like was supposed to be modelled after
> array-like in numpy. You don't see this in numpy docstrings:
>
> parameter : array_like or list
>
> I would agree that being explicit and adding string makes it easier to use
> for user not familiar with joblib.Memory (in general the docstring does say
> what happens if a string is passed in though). I was hoping that just using
> joblib.Memory-like in the type info and either adding a link to
> check_memory function for more details or copy and pasting the same
> docstring around would work.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9584#issuecomment-325892000>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz687XUZJa0BndNx5cSvAJLa49DacSks5sdP0QgaJpZM4O7xaJ>
> .
>

lesteve commented:
> Well, we can debate it, but I strongly disagree that "joblib.Memory-like" is sufficiently helpful to a user who's never heard of joblib.

Makes sense but then I would change the wording in the check_memory docstring, probably so that joblib.Memory-like means "same interface as joblib.Memory".

glemaitre commented:
Such ambiguity could be (partially) solved with a glossary as proposed in
one of the PR of Joel.
It would probably required to generate automatic link to redirect from
instance type to glossary
if the goal is to lower the barrier for new comer.
